doc/etsf_io.1
.\" It was generated by help2man 1.40.4.
.\" then modified by Damien Caliste.
.TH ETSF_IO "1" "November 2011" "etsf_io" "User Commands"
.SH NAME
ETSF_IO library \- manual page for etsf_io command line tool

.SH SYNOPSIS
.B etsf_io
[\fI-h | -a action\fR] [[\fI-i file\fR]...] [[\fI-f flag\fR]...]
[\-o file] [arguments]

.SH DESCRIPTION
.PP
Handle ETSF files, see \fB\-\-action\fR option.
.PP
\fB\-h\fR \fB\-\-help\fR             : show this little help.
.PP
\fB\-a\fR \fB\-\-action\fR value     : give the action to perform.
.PP
Possible action may be:
.PP
         * 'merge' to gather several files that have been splitted.
.PP
         * 'content' to get the name of specifications the file matches.
.PP
         * 'check' to check the validity of the file against specifications.
.PP
\fB\-o\fR \fB\-\-output\-file\fR file : give the path to the output ETSF file.
.PP
\fB\-i\fR \fB\-\-input\-file\fR file  : give the path for an input file.
                        This option can be used one or several times.
.PP
\fB\-l\fR \fB\-\-list\fR             : when action is check, it give the list of available flags.
.PP
\fB\-f\fR \fB\-\-flag\fR value       : give a flag name (get valid names from \fB\-l\fR option).

.SH EXAMPLES
Merge three files, etsf_io \fB\-a\fR merge \fB\-i\fR file1.nc \fB\-i\fR file2.nc
\fB\-i\fR file3.nc \fB\-o\fR output.nc
.PP
Get the contents of file test.nc, etsf_io \fB\-a\fR content test.nc
.PP
Get the list of flags for validity checks, etsf_io \fB\-a\fR check \fB\-l\fR
.PP
Checks with two flags, etsf_io \fB\-a\fR check \fB\-f\fR flag1 \fB\-f\fR flag2 test.nc

.SH AUTHOR
ETSF_IO library was written by Damien Caliste, Valerio Oleva, Yann Pouillon
and Matthieu Verstaete.
.PP
This manual page was written by Damien Caliste <damien.caliste@cea.fr>.



ChangeLog
* Thu Sep 19 2013 Yann Pouillon <yann.pouillon@ehu.es> and Damien Caliste <damien.caliste@cea.fr> 1.0.4
- Upgrade configure.ac to correct different badly written autoconf tests.
- Modify NetCDF detection for enhanced results.
- Add support for parallel I/O introduced in NetCDF 4.

* Mon Mar 29 2010 Damien Caliste <damien.caliste@cea.fr> 1.0.3
- Correct configure.ac and Makefile.am for aclocal to always run with its
  config/m4 inclusion.
- Modify the dielectric function specification to fullfil specs 3.3.
- Comment out the phonon part since not yet included officially in the
  spec.
- Modify the NetCDF detection in configure.ac to automatically detect
  NetCDF on Debian (using -lnetcdff and -lnetcdf).
- Modify the build system in doc to properly install the developper doc.

* Thu Nov 07 2008 Damien Caliste <damien.caliste@cea.fr> 1.0.3-dev2
- Add the proposal specification for the phonons.
- Modify the code generation to correctly handle the case with twice the
  same dimension in the shape declaration of an array. In that case _%d is
  appended to the name of the attricute in the group structure.
- Correct a bug to actually use the restriction on the number of elements
  for the variables that have a max_something dimension but that has
  no split flag.

* Thu Nov 06 2008 Damien Caliste <damien.caliste@cea.fr> 1.0.3-dev1
- Add the proposal specification for the dielectric function.

* Thu Sep 11 2008 Damien Caliste <damien.caliste@cea.fr> 1.0.2
- Update the fortran.m4 macro to recognise the open64 compiler.
- Update the configure.ac to handle the open64 compiler.

* Fri Apr 04 2008 Damien Caliste <damien.caliste@cea.fr> 1.0.1
- Add a --with-moduledir option to the configure to be able to choose the
	installation path for modules.
- Update the fortran.m4 macro to recognise newest version of gfortran (>4.1).

* Tue Aug 21 2007 Damien Caliste <damien.caliste@cea.fr> 1.0.0
- Official publication of version 1.0.0.

* Thu Aug 09 2007 Damien Caliste <damien.caliste@cea.fr> 1.0-rc2
- correct a bug in etsf_io when checking the wavefunctions in plane_waves
	mode with k_dependent = "no".
- add a simple routine in etsf_io_low_level to read the flag attribute as
	defined in the specifications. Also add tests for this routine.
- add the support of split definition in the check routines.
- bugs corrections in tests/low_level, src/low_level/read_routines.f90,
	the documentation...

* Fri Jul 06 2007 Damien Caliste <damien.caliste@cea.fr> 1.0-rc1
- add tests for the copy action of the ETSF groups.
- add a tutorial on the reading part, very simple with the basics of the
	etsf_io_<group>_get routines and the possible split definitions.
- add gestion of the use_time_reversal_at_gamma attribute as introduced by
	the 2.2 specifications.
- add a full backtrace for the routines in the library to ease the
	tracking of errors when called from an other program.
- add a way to precisely choose the variables when defining a file. This
	has bringed some minor in the ETSF group API, for etsf_io_main_*
	and for etsf_io_data_*.
- add the wavelet basis set as discussed in the IT9 meeting in San
	Sebastian in May 2007 (still in a draft for inclusion in
	specifications when 0.10 has been shipped).
	
* Fri May 11 2007 Damien Caliste <damien.caliste@cea.fr> 0.9
- complete all TODO occurances in the HTML pages.
- update all licence citation and remove all ABINIT occurance. Also add 2007 to the
  copyright and the name of the author for each part of the code.
- implement a copy_all_attribute routine in the low_level module. This routine
  is then used by the group level copy routines and the utils level merge routine
  to keep tracks of attributes in file merging. Several tests have been added
  in the low level pool to ensure this new copy routine.
- add a tutorial on the creation of non-ETSF variables, concurrently with ETSF
  ones.
- change the name of the wavedata group to basisdata for a better consistency.
	
* Fri Mar 09 2007 Damien Caliste <damien.caliste@cea.fr> 0.8
- add two more tutorials, one on the split usage and one on the high level routines.
- improve the documentation, especially the non robodoc pages.
- create a high level routine to check the validity of a file on the big domains
  the specifications deal with.
- create a high level routine to get the names of chemical species following the
  recommendations in the specifications.
- add two more actions to the etsf_io tool: the content and the check actions.
- fix some bugs in the build system and in the low level part.
	
* Wed Jan 17 2007 Damien Caliste <damien.caliste@cea.fr> 0.7
- add a new Fortran type to store the split definitions (e.g. my_kpoints...). This
  structure can be used when defining the variables on initialisation to set the
  splitted dimensions to the variables that need them.
- add a new source directory src/utils that will contains some high level library
  not mandatory but with convienient tools. It currently contains a routine to merge
  files that have been created with some split definition. To do it several group
  level have been added, especially some to copy variables from one file to another.
  The split action is complete for all ETSF variables, but other ones are currently
  lost, but also attributes and headers are lost.
- add a new tutorial on sub access. It explain how to use the
  <varname>__[spin|kpoint]_access attributes of groups that have been introduced in
  the last release.
	
* Tue Dec 12 2006 Damien Caliste <caliste@pcpm.ucl.ac.be> 0.6
- rewrite the generating script for low access in Python for a better portability.
- complete the implementation of all attributes (units, k_dependent and symmorphic).
  The unit attributes change the reading behavior, getting the values into atomic
  units, except if the optional argument use_atomic_units is set to false. The
  k_dependent attribute change the size of the array reduced_coordinates_of_plane_waves

src/utils/etsf_io.f90
program etsf_io_ploumploum

  use etsf_io_low_level
  use etsf_io
  use etsf_io_file

  implicit none

  integer :: nArg, iargc, i_arg
  character(len = 256) :: opt_value

  logical :: lstat
  type(etsf_io_low_error) :: error
  integer :: read_flags, check_flags, i, j
  type(etsf_io_low_error) :: errors(etsf_nspecs_data)
  
  character(len = 80) :: action = ""
  ! Variables used for the merge action
  logical :: get_help = .false.
  integer :: n_input_args = 0
  character(len = 256) :: input_args(256)
  integer :: n_input_files = 0
  character(len = 256) :: input_files(256)
  character(len = 256) :: output_file = ""
  logical :: get_specs_list = .false.
  integer :: n_input_flags = 0
  character(len = 256) :: input_flags(256)
  
  
  nArg = iargc()
  i_arg = 1
  do
     if (get_option("a", "action", opt_value, i_arg)) then
        action = opt_value(1:80)
     else if (get_option("i", "input-file", opt_value, i_arg)) then
        n_input_files = n_input_files + 1
        input_files(n_input_files) = opt_value
     else if (get_option("o", "output-file", opt_value, i_arg)) then
        output_file = opt_value
     else if (get_option("l", "list", opt_value, i_arg, .false.)) then
        get_specs_list = .true.
     else if (get_option("f", "flag", opt_value, i_arg)) then
        n_input_flags = n_input_flags + 1
        input_flags(n_input_flags) = opt_value
     else if (get_option("h", "help", opt_value, i_arg, .false.)) then
        get_help = .true.
     else
        ! This should be the end of option.
        call getarg(i_arg, opt_value)
        if (opt_value(1:2) == "-") then
           write(*, "(A)") "Error: unknown option or argument '", trim(opt_value), "'"
           call usage()
           stop 1
        else
           ! Store remaining argments in an array.
           n_input_args = 0
           do
              call getarg(i_arg, opt_value)
              n_input_args = n_input_args + 1
              write(input_args(n_input_args), "(A)") adjustl(opt_value)
              i_arg = i_arg + 1
              if (i_arg > narg) then
                 exit
              end if
           end do
        end if
     end if
     i_arg = i_arg + 1
     if (i_arg > narg) then
        exit
     end if
  end do

  if (get_help) then
     call usage()
     stop
  end if

  if (trim(action) == "merge") then
     if (n_input_files < 2) then
        write(*, "(A)") "Error: not enough input files for action merge."
        call usage()
        stop
     end if
     if (trim(output_file) == "") then
        write(*, "(A)") "Error: no output file for action merge."
        call usage()
        stop
     end if
!     write(*,*) "output file: ", trim(output_file)
     call etsf_io_file_merge(output_file, input_files(1:n_input_files), &
          & lstat, error)
     if (.not. lstat) then
        call etsf_io_low_error_handle(error)
        stop
     end if
  else if (trim(action) == "content") then
     if (n_input_args /= 1) then
        write(*, "(A)") "Error: exactly one arguments is required for action content."
        call usage()
        stop
     end if
     write(*,"(3A)") "Analyse file '", trim(input_args(1)), "'"
     call etsf_io_file_contents(read_flags, errors, trim(input_args(1)), &
          & lstat, error)
     if (.not. lstat) then
        call etsf_io_low_error_handle(error)
        stop
     end if
     do i = 1, etsf_nspecs_data, 1
        if (iand(read_flags, 2 ** (i - 1)) /= 0) then
           write(*,"(3A)") " - Ok - ", trim(etsf_specs_names(i)), "."
        else
           write(*,"(3A)") " - No - ", trim(etsf_specs_names(i)), "."
           write(*,"(4A)") "        given reason, '", trim(errors(i)%target_name), &
                & "' -> ", trim(errors(i)%error_message)
        end if
     end do
  else if (trim(action) == "check") then
     if (get_specs_list) then
        write(*,"(A)") "Available flags for specification checkings:"
        do i = 1, etsf_nspecs_data, 1
           write(*,"(A)") trim(etsf_specs_names(i))
        end do
        stop 1
     end if
     if (n_input_flags < 1) then
        write(*, "(A)") "Error: not enough flags for action check."
        call usage()
        stop 1
     end if
     if (n_input_args /= 1) then
        write(*, "(A)") "Error: exactly one arguments is required for action check."
        call usage()
        stop 1
     end if
     check_flags = 0
     do j = 1, n_input_args, 1
        do i = 1, etsf_nspecs_data, 1
           if (trim(etsf_specs_names(i)) == trim(input_flags(j))) then
              check_flags = check_flags + 2 ** (i - 1)
              exit
           end if
        end do
     end do
     call etsf_io_file_check(trim(input_args(1)), check_flags, lstat, error)
     if (.not. lstat) then
        call etsf_io_low_error_handle(error)
        stop 1
     end if
  else
     write(*, "(A)") "Error: missing or unknown action, use -a option."
     call usage()
     stop 1
  end if
  
contains

  function get_option(code, name, value, i_arg, with_value)
    implicit none

    character(len = 1), intent(in) :: code
    character(len = *), intent(in) :: name
    character(len = 256), intent(out) :: value
    integer, intent(inout) :: i_arg
    logical, intent(in), optional :: with_value
    logical :: get_option, my_with_value
    character(len = 256) :: arg_value
    integer :: start

    call getarg(i_arg, arg_value)
    if (arg_value(1:2) == "-"//code) then
       start = 3
       get_option = .true.
    else if (arg_value(1:len(name) + 2) == "--"//name) then
       start = len(name) + 3 
       get_option = .true.
    else
       get_option = .false.
    end if
    if (present(with_value)) then
       my_with_value = with_value
    else
       my_with_value = .true.
    end if
    if (get_option .and. my_with_value) then
       if (arg_value(start:start) /= " ") then
          if (arg_value(start:start) == "=") then
             value = arg_value(start + 1:256)
          else
             value = arg_value(start:256)
          end if
       else
          i_arg = i_arg + 1
          call getarg(i_arg, arg_value)
          value = adjustl(arg_value)
       end if
    end if
  end function get_option

  subroutine usage()
    write(*, "(A)") ""
    write(*, "(A)") "Usage: etsf_io [-h | -a action] [[-i file]...] [[-f flag]...]"
    write(*, "(A)") "               [-o file] [arguments]"
    write(*, "(A)") ""
    write(*, "(A)") "   Handle ETSF files, see --action option."
    write(*, "(A)") "-h --help             : show this little help."
    write(*, "(A)") "-a --action value     : give the action to perform."
    write(*, "(A)") "                        Possible action may be:"
    write(*, "(A)") "                        * 'merge' to gather several files that"
    write(*, "(A)") "                          have been splitted."
    write(*, "(A)") "                        * 'content' to get the name of"
    write(*, "(A)") "                          specifications the file matches."
    write(*, "(A)") "                        * 'check' to check the validity of"
    write(*, "(A)") "                          the file against specifications."
    write(*, "(A)") "-o --output-file file : give the path to the output ETSF file."
    write(*, "(A)") "-i --input-file file  : give the path for an input file. This"
    write(*, "(A)") "                        option can be used one or several times."
    write(*, "(A)") "-l --list             : when action is check, it give the list"
    write(*, "(A)") "                        of available flags."
    write(*, "(A)") "-f --flag value       : give a flag name (get valid names from"
    write(*, "(A)") "                        -l option)."
    write(*, "(A)") ""
    write(*, "(A)") "   Examples:"
    write(*, "(A)") "Merge three files, etsf_io -a merge -i file1.nc -i file2.nc"
    write(*, "(A)") "                   -i file3.nc -o output.nc"
    write(*, "(A)") ""
    write(*, "(A)") "Get the contents of file test.nc, etsf_io -a content test.nc"
    write(*, "(A)") ""
    write(*, "(A)") "Get the list of flags for validity checks, etsf_io -a check -l"
    write(*, "(A)") ""
    write(*, "(A)") "Checks with two flags, etsf_io -a check -f flag1 -f flag2 test.nc"
  end subroutine usage

end program etsf_io_ploumploum
